trestle/core/rules_validator.py

```diff
@@ -0,0 +1,110 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Validate by confirming rule parameter values are consistent."""
+import logging
+import pathlib
+
+import trestle.oscal.catalog as cat
+from trestle.common.common_types import TopLevelOscalModel
+from trestle.common.common_types import TypeWithProps
+from trestle.common.list_utils import as_list, deep_get, deep_set, get_item_from_list
+from trestle.core.catalog.catalog_interface import CatalogInterface
+from trestle.core.profile_resolver import ProfileResolver
+from trestle.core.validator import Validator
+from trestle.oscal.ssp import SystemSecurityPlan
+
+logger = logging.getLogger(__name__)
+
+shared_params = {}
+
+
+class RulesValidator(Validator):
+    """Validator to confirm all rule parameter values are consistent."""
+
+    def add_shared_param(
+        self, item: TypeWithProps, control_id: str, cat_int: CatalogInterface, comp_name: str = ''
+    ) -> None:
+        """
+        Add a rule shared parameter to the rule shared parameters list.
+
+        args:
+            item: Generic item to iterate over parameters.
+            control_id: Current control id.
+            cat_int: Instance of catalog interface with controls catalog loaded.
+            comp_name: Component name to save.
+        """
+        for param in as_list(item.set_parameters if not isinstance(item, cat.Control) else item.params):
+            parameter_id = param.param_id if not isinstance(item, cat.Control) else param.id
+            control = cat_int.get_control_by_param_id(parameter_id)
+            if not control:
+                deep_set(shared_params, [parameter_id, comp_name, control_id], param.values)
+
+    def model_is_valid(self, model: TopLevelOscalModel, quiet: bool, trestle_root: pathlib.Path = '') -> bool:
+        """
+        Test if the model is valid.
+
+        args:
+            model: A top level OSCAL model.
+            quiet: Don't report msgs unless invalid.
+            trestle_root: Trestle root path.
+
+        returns:
+            True (valid) if the model's rule parameter values are the same accross controls.
+        """
+        # verify if model type is either an SSP of a Component Definition
+        if not isinstance(model, (SystemSecurityPlan)):
+            return True
+
+        components = as_list(model.system_implementation.components)
+        component_titles = [x.title for x in components]
+        profile_resolver = ProfileResolver()
+        catalog = profile_resolver.get_resolved_profile_catalog(trestle_root, model.import_profile.href)
+        catalog_interface = CatalogInterface(catalog)
+
+        for control in catalog_interface.get_all_controls_from_catalog(True):
+            self.add_shared_param(control, control.id, catalog_interface)
```

Comment:
no need for this.  when you add the param, check to see if it is known in the catalog interface by param_id.  If it is known then ignore it and do not add it to the dict.

Reply:
got it.